cmd/default-domain/main.go

```diff
@@ -50,89 +57,115 @@ const (
+								ServicePort:      intstr.FromInt(80),
+							},
+						}},
+					}},
+				},
+			}},
+		},
+	})
 	if err != nil {
-		return nil, fmt.Errorf("error looking up IngressGateway: %w", err)
-	}
-	// Walk the list of Ingress entries in the Service's LoadBalancer status.
-	// If one of IP address is found, return it.  Otherwise, keep one with
-	// a hostname assigned, if any.
-	var hostname *corev1.LoadBalancerIngress
-	for i, ing := range svc.Status.LoadBalancer.Ingress {
-		if ing.IP != "" {
-			return &svc.Status.LoadBalancer.Ingress[i], nil
-		}
-		if ing.Hostname != "" {
-			hostname = &svc.Status.LoadBalancer.Ingress[i]
+		return nil, err
+	}
+	defer client.NetworkingV1alpha1().Ingresses(system.Namespace()).Delete(ing.Name, &metav1.DeleteOptions{})
+
+	// Wait for the Ingress to be Ready.
+	if err := wait.PollImmediate(pollInterval, waitTimeout, func() (done bool, err error) {
+		ing, err = client.NetworkingV1alpha1().Ingresses(system.Namespace()).Get(
+			ing.Name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
 		}
+		return ing.Status.IsReady(), nil
+	}); err != nil {
+		return nil, err
 	}
-	if hostname != nil {
-		return hostname, nil
+	if len(ing.Status.PublicLoadBalancer.Ingress) == 0 {
+		return nil, errors.New("Ingress has no public load balancers in status")
 	}
-	return nil, fmt.Errorf("service %s/%s does not have an assigned external address",
-		svc.Namespace, svc.Name)
-}
 
-func waitForIngressGatewayAddress(kubeclient *kubernetes.Clientset) (addr *corev1.LoadBalancerIngress, waitErr error) {
-	logger := logging.FromContext(context.Background()).Named(appName)
-	waitErr = wait.PollImmediate(pollInterval, waitTimeout, func() (done bool, err error) {
-		addr, err = lookupIngressGatewayAddress(kubeclient)
-		if err == nil {
-			return true, nil
+	// We expect an ingress LB with the form foo.bar.svc.cluster.local (though
+	// we aren't strictly sensitive to the suffix, this is just illustrative).
+	internalDomain := ing.Status.PublicLoadBalancer.Ingress[0].DomainInternal
+	parts := strings.SplitN(internalDomain, ".", 3)
+	if len(parts) < 3 {
+		return nil, fmt.Errorf("Ingress public load balancer had unexpected shape: %q", internalDomain)
+	}
+	name, namespace := parts[0], parts[1]
+
+	// Wait for the Ingress Service to have an external IP.
+	var svc *corev1.Service
+	if err := wait.PollImmediate(pollInterval, waitTimeout, func() (done bool, err error) {
+		svc, err = kubeclient.CoreV1().Services(namespace).Get(name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
 		}
-		logger.Infof("Failed lookup IngressGateway, retrying: %v", err)
-		return false, nil
-	})
-	return addr, waitErr
+		if len(svc.Status.LoadBalancer.Ingress) == 0 {
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		return nil, err
+	}
+	return &svc.Status.LoadBalancer.Ingress[0], nil
 }
 
 func main() {
 	flag.Parse()
 	logger := logging.FromContext(context.Background()).Named(appName)
```

Comment:
Truly, I am not sure this returns any useful logger, given we don't read CMs here. I'd just go with klog/glog here.

Reply:
Ack, this is pre-existing, so I'm not inclined to change things.